sympy/polys/sqfreetools.py

```diff
@@ -401,30 +401,36 @@ def dmp_sqf_list(f, u, K, all=False):
     deg = dmp_degree(f, u)
     if deg < 0:
         return coeff, []
-    elif deg == 0:
-        coeff2, factors = dmp_sqf_list(dmp_LC(f, u), u-1, K, all=all)
-        factors = [([fac], exp) for fac, exp in factors]
-        return coeff*coeff2, factors
 
-    result, i = [], 1
+    content, f = dmp_primitive(f, u, K)
 
-    h = dmp_diff(f, 1, u, K)
-    g, p, q = dmp_inner_gcd(f, h, u, K)
+    result = []
 
-    while True:
-        d = dmp_diff(p, 1, u, K)
-        h = dmp_sub(q, d, u, K)
+    if deg != 0:
 
-        if dmp_zero_p(h, u):
-            result.append((p, i))
-            break
+        h = dmp_diff(f, 1, u, K)
+        g, p, q = dmp_inner_gcd(f, h, u, K)
 
-        g, p, q = dmp_inner_gcd(p, h, u, K)
+        i = 1
 
-        if all or dmp_degree(g, u) > 0:
-            result.append((g, i))
+        while True:
+            d = dmp_diff(p, 1, u, K)
+            h = dmp_sub(q, d, u, K)
 
-        i += 1
+            if dmp_zero_p(h, u):
+                result.append((p, i))
+                break
+
+            g, p, q = dmp_inner_gcd(p, h, u, K)
+
+            if all or dmp_degree(g, u) > 0:
+                result.append((g, i))
```

Comment:
The idea here is that we have a polynomial `p = p1 * p2**2 * p3**3 * ...` and we want to find `p1`, `p2`, etc which are coprime. The way we do that is by differentiating the polynomial and computing the gcd like `g = gcd(p, p')`. This works because if `p2**2` divides `p` then `p2` must divide both `p` and `p'` (you can check by the product rule). Therefore `p2` must divide `g`. However `p2` will not divide `p''`. We can find `p2` then by observing that it is a factor of `p` and `p'` but not of `p''`. What matters for the multiplicity of the factor is not the degree of `g` or the degree of `p2` but rather how many times we have differentiated i.e. `i`.

The actual (Yun's) algorithm is a little more complicated than this but it also described on Wikipedia:
https://en.wikipedia.org/wiki/Square-free_polynomial#Yun's_algorithm

What is not mentioned there but is mentioned in Yun's paper is that for multivariate polynomials like `Q[x,y,z]` we need to think of them as being like `p(z) in Q[x,y][z]` where the coefficients of `p(z)` (elements of `Q[x,y]`) should be coprime. In other words `p(z)` should be primitive in the sense of a univariate polynomial being primitive. This is the difference between `dmp_ground_primitive` (make primitive in `Q[x,y,z]`) and `dmp_primitive` (make primitive in `Q[x,y][z]`). With `dmp_ground_primitive` the extracted content is an element of `Q` but with `dmp_primitive` it is an element of `Q[x,y]`. The change I made here is to extract that content in `Q[x,y]` and then recursively compute its square free factorisation.

What is potentially worth changing is that this version of `dmp_sqf_list` does not produce the factors in order whereas the previous code does:
```python
In [1]: R, x, y = ring(['x', 'y'], QQ)

In [2]: p = (x+y)**2*(x-1)*(y-1)*(x-2)**2*(y-2)**2

In [3]: R.dmp_sqf_list(p)
Out[3]: (MPQ(1,1), [(x - 1, 1), (x**2 + x*y - 2*x - 2*y, 2), (y - 1, 1), (y - 2, 2)])
```
With master that would be:
```python
In [5]: R.dmp_sqf_list(p)
Out[5]: (MPQ(1,1), [(x - 1, 1), (x**2 + x*y - 2*x - 2*y, 2)])
```
The master result is incorrect but it always has a well define format with ascending powers because results are appended as `i` increases.

The higher level `sqf` function somehow accommodates for this but `Poly.sqf` does not:
```python
In [1]: p = (x+y)**2*(x-1)*(y-1)*(x-2)**2*(y-2)**2

In [2]: sqf(expand(p))
Out[2]:
                                                                 2
                  ⎛ 2        2      2                    2      ⎞
(x⋅y - x - y + 1)⋅⎝x ⋅y - 2⋅x  + x⋅y  - 4⋅x⋅y + 4⋅x - 2⋅y  + 4⋅y⎠

In [4]: Poly(p).sqf_list()
Out[4]: (1, [(Poly(x - 1, x, y, domain='ZZ'), 1), (Poly(x**2 + x*y - 2*x - 2*y, x, y, domain='ZZ'), 2), (Poly(y - 1, x, y, domain='ZZ'), 1), (Poly(y - 2, x, y, domain='ZZ'), 2)])
```
I'll patch this up in `dmp_sqf_list`.